internal/ethereum/confirmation_reconciler_test.go

```
@@ -967,6 +913,8 @@ func TestCompareAndUpdateConfirmationQueue_AlreadyConfirmableConnectable(t *test
 }
 
 func TestCompareAndUpdateConfirmationQueue_AlreadyConfirmableButAllExistingConfirmationsAreTooHighForTargetConfirmationCount(t *testing.T) {
+	t.Skip("Skipping test because I don't understand this behavior") //Trying to understand the reasoning behind supporting an invalid list of existing confirmations.  Should this not just be an error?
```

Good question. It's clearly an error; the question is why it's valuable to ask the user to handle the error, other than telling them the current correct confirmation based on the latest chain state.

---

I did consider whether it would be better to just throw away any invalid existing confirmations and continue into the algorithm as if the user hadn't provided any existing confirmations ( or as if they had just provided the valid subset of them).  But my worry there is that if the users code is working correctly, then these should never be invalid because they are just giving us back what we returned to them on a previous iteration.  Sure, the chain might have been re-organized since then, but the linked list should be consistent within itself.  If it is not, then that is an indication that either the user's code has a bug or there memory / storage has been corrupted and it seemed to me that they would need an error in that case.

---

@hosie After our chat, I get the point that it would be nice to surface confirmation queue corruption to the client if they have been editing the confirmation list.

However, the client should not edit the confirmation list at all. If they don't want the efficiency, they don't need to pass over historical output at all. They can just call the function with no existing confirmations provided.

Therefore, a corrupted confirmation list is treated as a broken output by a previous version of this function. So the function should clean up its own mess without user intervention by default. 

 I'm still of the opinion that the reconcile function should just return the correct confirmation

We can introduce a new flag to highlight that the input is corrupted if needed. I had a `rebuilt` flag initially, but removed it later because when I write the logic on the other side, the logic didn't care. It's passing around the whole output as a blackbox, `confirmed` is the one that matters for confirmation logic, the details are readonly information for users to view the progress, therefore treated as a blackbox.